src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -106,35 +106,45 @@ private static unsafe SafeFileHandle CreateFile(string fullPath, FileMode mode,
             return fileHandle;
         }
 
-        private static unsafe void Preallocate(string fullPath, long preallocationSize, SafeFileHandle fileHandle)
+        // preallocationSize must be ignored for non-seekable files, unsupported file systems
+        // and other failures other than ERROR_DISK_FULL and ERROR_FILE_TOO_LARGE
+        private static unsafe void Preallocate(string fullPath, long preallocationSize, SafeFileHandle fileHandle, FileMode mode)
         {
-            var allocationInfo = new Interop.Kernel32.FILE_ALLOCATION_INFO
+            if (mode == FileMode.OpenOrCreate)
             {
-                AllocationSize = preallocationSize
+                Interop.Kernel32.FILE_STANDARD_INFO info;
+                if (!Interop.Kernel32.GetFileInformationByHandleEx(fileHandle, Interop.Kernel32.FileStandardInfo, &info, (uint)sizeof(Interop.Kernel32.FILE_STANDARD_INFO))
+                    || info.EndOfFile >= preallocationSize) // existing files must NOT be shrinked
+                {
+                    return;
+                }
+            }
+
+            var eofInfo = new Interop.Kernel32.FILE_END_OF_FILE_INFO
+            {
+                EndOfFile = preallocationSize
             };
 
             if (!Interop.Kernel32.SetFileInformationByHandle(
                 fileHandle,
-                Interop.Kernel32.FileAllocationInfo,
-                &allocationInfo,
-                (uint)sizeof(Interop.Kernel32.FILE_ALLOCATION_INFO)))
+                Interop.Kernel32.FileEndOfFileInfo,
+                &eofInfo,
+                (uint)sizeof(Interop.Kernel32.FILE_END_OF_FILE_INFO)))
             {
                 int errorCode = Marshal.GetLastPInvokeError();
+                if (errorCode != Interop.Errors.ERROR_DISK_FULL && errorCode != Interop.Errors.ERROR_FILE_TOO_LARGE)
+                {
+                    return;
+                }
```

Comment:
After this, the only benefit on Windows of using preallocation size rather than just calling SetLength is that some error conditions are ignored?

These errors should be exceedingly rare.  Rather than duplicating all this code, should we just do:
```C#
try
{
    SetLength(...);
}
catch { ... /* whatever special you want to do */ ... }
```
?  Or somehow share the code in another way?